File: sdk-java/src/main/java/io/littlehorse/sdk/worker/WorkerContext.java
package io.littlehorse.sdk.worker;

import io.littlehorse.sdk.common.LHLibUtil;
import io.littlehorse.sdk.common.proto.Checkpoint;
import io.littlehorse.sdk.common.proto.CheckpointId;
import io.littlehorse.sdk.common.proto.LittleHorseGrpc.LittleHorseBlockingStub;
import io.littlehorse.sdk.common.proto.NodeRunId;
import io.littlehorse.sdk.common.proto.PutCheckpointRequest;
import io.littlehorse.sdk.common.proto.PutCheckpointResponse;
import io.littlehorse.sdk.common.proto.PutCheckpointResponse.FlowControlContinue;
import io.littlehorse.sdk.common.proto.ScheduledTask;
import io.littlehorse.sdk.common.proto.TaskRunId;
import io.littlehorse.sdk.common.proto.TaskRunSource;
import io.littlehorse.sdk.common.proto.UserTaskTriggerReference;
import io.littlehorse.sdk.common.proto.WfRunId;
import java.util.Date;

/**
 * This class contains runtime information about the specific WfRun and NodeRun that is being
 * executed by the Task Worker. It may optionally be added into the input parameters of your
 * LHTaskMethod, and the Runtime will provision the WorkerContext and pass it into the method.
 */
public class WorkerContext {

    private ScheduledTask scheduledTask;

    private Date scheduleTime;
    private String logOutput;
    private int checkpointsSoFarInThisRun;
    private LittleHorseBlockingStub client;

    /**
     * Constructor for internal use by the Task Worker Library.
     *
     * @param scheduledTask is the raw payload for the scheduled task.
     * @param scheduleTime is the time that the task was actually scheduled.
     */
    public WorkerContext(ScheduledTask scheduledTask, Date scheduleTime, LittleHorseBlockingStub client) {
        this.scheduledTask = scheduledTask;
        this.scheduleTime = scheduleTime;
        this.logOutput = "";
        checkpointsSoFarInThisRun = 0;
        this.client = client;
    }

    /**
     * Returns the Id of the WfRun for the NodeRun that's being executed.
     *
     * @return the Id of the WfRun for the NodeRun that's being executed.
     */
    public WfRunId getWfRunId() {
        return LHLibUtil.getWfRunId(scheduledTask.getSource());
    }

    /**
     * Returns the NodeRun ID for the Task that was just scheduled.
     *
     * @return a `NodeRunIdPb` protobuf class with the ID from the executed NodeRun.
     */
    public NodeRunId getNodeRunId() {
        TaskRunSource source = scheduledTask.getSource();
        switch (source.getTaskRunSourceCase()) {
            case TASK_NODE:
                return source.getTaskNode().getNodeRunId();
            case USER_TASK_TRIGGER:
                return source.getUserTaskTrigger().getNodeRunId();
            case TASKRUNSOURCE_NOT_SET:
        }
        return null;
    }

    /**
     * Returns the attemptNumber of the NodeRun that's being executed. If this is the first attempt,
     * returns zero. If this is the first retry, returns 1, and so on.
     *
     * @return the attempt number of the NodeRun that's being executed.
     */
    public int getAttemptNumber() {
        return scheduledTask.getAttemptNumber();
    }

    /**
     * Returns the time at which the task was scheduled by the processor. May be useful in certain
     * customer edge cases, eg. to determine whether it's too late to actually perform an action,
     * when (now() - getScheduledTime()) is above some threshold, etc.
     *
     * @return the time at which the current NodeRun was scheduled.
     */
    public Date getScheduledTime() {
        return scheduleTime;
    }

    /**
     * Provides a way to push data into the log output. Any object may be passed in; its String
     * representation will be appended to the logOutput of this NodeRun.
     *
     * @param thing the Object to log to the NodeRun's logOutput.
     */
    public void log(Object thing) {
        if (thing != null) {
            logOutput += thing.toString();
        } else {
            logOutput += "null";
        }
    }

    /**
     * Returns the current logOutput.
     *
     * @return the current log output.
     */
    public String getLogOutput() {
        return logOutput;
    }

    /**
     * Returns the TaskRunId of this TaskRun.
     * @return the associated TaskRunId.
     */
    public TaskRunId getTaskRunId() {
        return scheduledTask.getTaskRunId();
    }

    private UserTaskTriggerReference getUserTaskTrigger() {
        return scheduledTask.getSource().hasUserTaskTrigger()
                ? scheduledTask.getSource().getUserTaskTrigger()
                : null;
    }

    /**
     * If this TaskRun is a User Task Reminder TaskRun, then this method returns the
     * UserId of the user who the associated UserTask is assigned to. Returns
     * null if:
     * - this TaskRun is not a Reminder Task
     * - this TaskRun is a Reminder Task, but the UserTaskRun does not have an assigned
     *   user id.
     *
     * @return the id of the user that the associated UserTask is assigned to.
     */
    public String getUserId() {
        UserTaskTriggerReference uttr = getUserTaskTrigger();
        if (uttr == null) return null;

        return uttr.hasUserId() ? uttr.getUserId() : null;
    }

    /**
     * If this TaskRun is a User Task Reminder TaskRun, then this method returns the
     * UserGroup that the associated UserTask is assigned to. Returns null if:
     * - this TaskRun is not a Reminder Task
     * - this TaskRun is a Reminder Task, but the UserTaskRun does not have an
     *   associated User Group
     *
     * @return the id of the User Group that the associated UserTask is assigned to.
     */
    public String getUserGroup() {
        UserTaskTriggerReference uttr = getUserTaskTrigger();
        if (uttr == null) return null;

        return uttr.hasUserGroup() ? uttr.getUserGroup() : null;
    }

    /**
     * Returns an idempotency key that can be used to make calls to upstream api's idempotent across
     * TaskRun Retries.
     *
     * @return an idempotency key.
     */
    public String getIdempotencyKey() {
        return LHLibUtil.taskRunIdToString(getTaskRunId());
    }

    public <T> T executeAndCheckpoint(CheckpointableFunction<T> runnable, Class<T> clazz) {
        if (checkpointsSoFarInThisRun < scheduledTask.getTotalObservedCheckpoints()) {
            return (T) fetchCheckpoint(checkpointsSoFarInThisRun++, clazz);
        } else {
            return saveCheckpoint(runnable, clazz);
        }
    }

    private <T> T fetchCheckpoint(int checkpointNumber, Class<T> clazz) {
        CheckpointId id = CheckpointId.newBuilder()
                .setTaskRun(scheduledTask.getTaskRunId())
                .setCheckpointNumber(checkpointNumber)
                .build();
        Checkpoint checkpoint = client.getCheckpoint(id);
        return (T) LHLibUtil.varValToObj(checkpoint.getValue(), clazz);
    }

    private <T> T saveCheckpoint(CheckpointableFunction<T> runnable, Class<T> clazz) {
        CheckpointContext checkpointContext = new CheckpointContext();
        T result = runnable.run(checkpointContext);

        PutCheckpointResponse response = client.putCheckpoint(PutCheckpointRequest.newBuilder()
                .setTaskAttempt(scheduledTask.getAttemptNumber())
                .setTaskRunId(scheduledTask.getTaskRunId())
                .setValue(LHLibUtil.objToVarVal(result))
                .setLogs(checkpointContext.getLogOutput())
                .build());

        checkpointsSoFarInThisRun++;

        if (response.getFlowControlContinueType() != FlowControlContinue.CONTINUE_TASK) {
            throw new RuntimeException("Halting execution because the server told us to.");
        }
        return result;
    }
}

File: sdk-js/src/worker/WorkerContext.ts
import { PutCheckpointResponse_FlowControlContinue, ScheduledTask } from '../proto/service'
import { Timestamp } from '../proto/google/protobuf/timestamp'
import { WfRunId, TaskRunId, NodeRunId } from '../proto/object_id'
import { UserTaskTriggerReference } from '../proto/user_tasks'
import type { LHPublicClient } from '../client'
import { extractVariableValue, toVariableValue } from './variableMapping'
import { LHTypeAdapterRegistry } from '../common/typeAdapters'

/**
 * Collects logs produced inside a checkpointed operation; they are stored
 * with the checkpoint rather than the TaskRun.
 */
export class CheckpointContext {
  private logOutput = ''

  log(message: string): void {
    if (this.logOutput.length > 0) this.logOutput += '\n'
    this.logOutput += message
  }

  getLogOutput(): string {
    return this.logOutput
  }
}

/**
 * Context object provided to task functions during execution. Contains metadata
 * about the current task run and allows logging.
 */
export class WorkerContext {
  private logOutput: string = ''
  private readonly task: ScheduledTask
  private readonly client?: LHPublicClient
  /** How many checkpoints this attempt has reached so far. */
  private checkpointsSoFar = 0

  constructor(
    task: ScheduledTask,
    client?: LHPublicClient,
    private readonly typeAdapters?: LHTypeAdapterRegistry
  ) {
    this.task = task
    this.client = client
  }

  /**
   * Returns the WfRunId of the WfRun that triggered this TaskRun.
   */
  getWfRunId(): WfRunId | undefined {
    return this.task.taskRunId?.wfRunId
  }

  /**
   * Returns the TaskRunId of the current TaskRun.
   */
  getTaskRunId(): TaskRunId | undefined {
    return this.task.taskRunId
  }

  /**
   * Returns the attempt number of the current TaskRun (0-indexed).
   * 0 means this is the first attempt; 1 means first retry, etc.
   */
  getAttemptNumber(): number {
    return this.task.attemptNumber
  }

  /**
   * Returns the time at which this TaskRun was scheduled.
   */
  getScheduledTime(): string | undefined {
    return this.task.createdAt ? Timestamp.toDate(this.task.createdAt).toISOString() : undefined
  }

  /**
   * Returns the NodeRunId if this task was triggered by a TASK node.
   */
  getNodeRunId(): NodeRunId | undefined {
    if (this.task.source?.taskRunSource?.oneofKind === 'taskNode') {
      return this.task.source.taskRunSource.taskNode.nodeRunId
    }
    return undefined
  }

  /**
   * Returns a deterministic idempotency key derived from the TaskRunId.
   */
  getIdempotencyKey(): string | undefined {
    const taskRunId = this.task.taskRunId
    if (!taskRunId) return undefined
    const wfRunId = taskRunId.wfRunId?.id ?? ''
    return `${wfRunId}/${taskRunId.taskGuid}`
  }

  /**
   * For a User Task reminder TaskRun, the user the UserTaskRun is assigned to.
   * Undefined when this is not a reminder task, or it has no assigned user.
   */
  getUserId(): string | undefined {
    return this.userTaskTrigger()?.userId
  }

  /**
   * For a User Task reminder TaskRun, the group the UserTaskRun is assigned
   * to. Undefined when this is not a reminder task, or it has no group.
   */
  getUserGroup(): string | undefined {
    return this.userTaskTrigger()?.userGroup
  }

  private userTaskTrigger(): UserTaskTriggerReference | undefined {
    const source = this.task.source?.taskRunSource
    return source?.oneofKind === 'userTaskTrigger' ? source.userTaskTrigger : undefined
  }

  /**
   * Runs `operation` once across all attempts of this TaskRun.
   *
   * On a retry the server reports how many checkpoints the previous attempt
   * reached (`totalObservedCheckpoints`); calls up to that count are replayed
   * from the stored value instead of being re-executed, so side effects that
   * already happened — charging a card, sending a message — are not repeated.
   * Mirrors Java's WorkerContext#executeAndCheckpoint.
   */
  async executeAndCheckpoint<T>(operation: (ctx: CheckpointContext) => T | Promise<T>): Promise<T> {
    if (this.client === undefined) {
      throw new Error('executeAndCheckpoint requires a server connection; none was provided to this WorkerContext')
    }

    if (this.checkpointsSoFar < this.task.totalObservedCheckpoints) {
      const checkpointNumber = this.checkpointsSoFar++
      const checkpoint = await this.client.getCheckpoint({
        taskRun: this.task.taskRunId,
        checkpointNumber,
      })
      return extractVariableValue(checkpoint.value) as T
    }

    const checkpointContext = new CheckpointContext()
    const result = await operation(checkpointContext)

    const response = await this.client.putCheckpoint({
      taskRunId: this.task.taskRunId,
      taskAttempt: this.task.attemptNumber,
      value: toVariableValue(result, undefined, this.typeAdapters),
      logs: checkpointContext.getLogOutput(),
    })
    this.checkpointsSoFar++

    if (response.flowControlContinueType !== PutCheckpointResponse_FlowControlContinue.CONTINUE_TASK) {
      throw new Error('Halting execution because the server told us to.')
    }
    return result
  }

  /**
   * Appends a log message. Logs are sent back to the server with the task result.
   */
  log(message: string): void {
    if (this.logOutput.length > 0) {
      this.logOutput += '\n'
    }
    this.logOutput += message
  }

  /**
   * Returns the accumulated log output.
   */
  getLogOutput(): string {
    return this.logOutput
  }
}
